lib/src/webvtt/WebVttParser.ts

```diff
@@ -0,0 +1,466 @@
+	 * A callback function that is called when a region is parsed.
+	 */
+	onregion?: (region: WebVttRegion) => void;
+
+	/**
+	 * A callback function that is called when a timestamp map is parsed.
+	 */
+	ontimestampmap?: (timestampMap: TimestampMap) => void;
+
+	/**
+	 * A callback function that is called when a cue is parsed.
+	 */
+	oncue?: (cue: WebVttCue) => void;
+
+	/**
+	 * A callback function that is called when a style is parsed.
+	 */
+	onstyle?: (style: string) => void;
+
+	/**
+	 * A callback function that is called when the parser is flushed.
+	 */
+	onflush?: () => void;
+
+	/**
+	 * Create a new WebVTT parser.
+	 *
+	 * @param options - The options to use for the parser.
+	 */
+	constructor(options: WebVttParserOptions = {}) {
+		const useDomTypes = options.useDomTypes ?? true;
+		this.createCue = options.createCue || useDomTypes ? createCue : defaultFactory<WebVttCue>;
+		this.createRegion = options.createRegion || useDomTypes ? createRegion : defaultFactory<WebVttRegion>;
+
+		this.state = WebVttParserState.INITIAL;
+		this.buffer = '';
+		this.style = '';
+		this.regionList = [];
+	}
+
+	/**
+	 * Parse the given data.
+	 *
+	 * @param data - The data to parse.
+	 * @param reuseCue - Whether to reuse the cue.
+	 * @returns The parser.
+	 */
+	parse(data?: string, reuseCue: boolean = false): WebVttParser {
+		// If there is no data then we will just try to parse whatever is in buffer already.
+		// This may occur in circumstances, for example when flush() is called.
+		if (data) {
+			this.buffer += data;
+		}
+
+		const collectNextLine = (): string => {
+			const buffer = this.buffer;
+			let pos = 0;
+			while (pos < buffer.length && buffer[pos] !== '\r' && buffer[pos] !== '\n') {
+				++pos;
+			}
+			const line = buffer.substr(0, pos);
+			// Advance the buffer early in case we fail below.
+			if (buffer[pos] === '\r') {
+				++pos;
+			}
+			if (buffer[pos] === '\n') {
+				++pos;
+			}
+			this.buffer = buffer.substr(pos);
+			return line;
+		};
+
+		// draft-pantos-http-live-streaming-20
+		// https://tools.ietf.org/html/draft-pantos-http-live-streaming-20#section-3.5
+		// 3.5 WebVTT
+		const parseTimestampMap = (input: string): void => {
+			const settings = new Settings();
+
+			parseOptions(input, (k: string, v: string): void => {
+				switch (k) {
+					case 'MPEGT':
+						settings.integer(k + 'S', v);
+						break;
+					case 'LOCA':
+						settings.set(k + 'L', parseTimeStamp(v));
+						break;
+				}
+			}, /[^\d]:/, /,/);
+
+			this.ontimestampmap?.({
+				'MPEGTS': settings.get('MPEGTS'),
+				'LOCAL': settings.get('LOCAL'),
+			});
+		};
+
+		// 3.2 WebVtt metadata header syntax
+		const parseHeader = (input: string): void => {
+			if (input.match(/X-TIMESTAMP-MAP/)) {
+				// This line contains HLS X-TIMESTAMP-MAP metadata
+				parseOptions(input, (k: string, v: string): void => {
+					switch (k) {
+						case 'X-TIMESTAMP-MAP':
+							parseTimestampMap(v);
+							break;
+					}
+				}, /=/);
+			}
+		};
```

Comment:
@gkatsev `X-TIMESTAMP-MAP` is mentioned in the HLS spec, but not the latest WebVTT spec. I don't see anything in the WebVTT spec that would account for this line, except maybe the mention of ["file magic" in section 4.1 (item #4)] (https://www.w3.org/TR/webvtt1/#file-structure). Do you know if the concept of ["metadata headers"](https://www.w3.org/TR/2015/WD-webvtt1-20151208/#webvtt-metadata-header) still exists?

Reply:
it's worth keeping around for hls support. In WebVTT, we want to bring back those headers as an extension point for HLS and others https://github.com/w3c/webvtt/issues/485#issuecomment-646089841, but it hasn't happened yet. Once it comes back to WebVTT the HLS spec will update to point it but would continue defining it as WebVTT would just say that there can be data in that location and not what it says.